File: protocol/src/test/kotlin/com/zfoo/kotlin/ProtocolManager.kt
package com.zfoo.kotlin
import com.zfoo.kotlin.packet.EmptyObject
import com.zfoo.kotlin.packet.ComplexObject
import com.zfoo.kotlin.packet.NormalObject
import com.zfoo.kotlin.packet.ObjectA
import com.zfoo.kotlin.packet.ObjectB
import com.zfoo.kotlin.packet.SimpleObject
class ProtocolManager {
    companion object {
        val protocols = arrayOfNulls<IProtocolRegistration>(Short.MAX_VALUE.toInt())
        var protocolIdMap: MutableMap<Class<*>, Short> = HashMap()
        @JvmStatic
        fun initProtocol() {
            // initProtocol
            protocols[0] = EmptyObject.registrationEmptyObject
            protocolIdMap[EmptyObject::class.java] = 0.toShort()
            protocols[100] = ComplexObject.registrationComplexObject
            protocolIdMap[ComplexObject::class.java] = 100.toShort()
            protocols[101] = NormalObject.registrationNormalObject
            protocolIdMap[NormalObject::class.java] = 101.toShort()
            protocols[102] = ObjectA.registrationObjectA
            protocolIdMap[ObjectA::class.java] = 102.toShort()
            protocols[103] = ObjectB.registrationObjectB
            protocolIdMap[ObjectB::class.java] = 103.toShort()
            protocols[104] = SimpleObject.registrationSimpleObject
            protocolIdMap[SimpleObject::class.java] = 104.toShort()
        }

        @JvmStatic
        fun getProtocolId(clazz: Class<*>): Short {
            return protocolIdMap[clazz]!!
        }

        @JvmStatic
        fun getProtocol(protocolId: Short): IProtocolRegistration {
            return protocols[protocolId.toInt()]
                ?: throw RuntimeException("[protocolId:$protocolId] not exist")
        }

        @JvmStatic
        fun write(buffer: ByteBuffer, packet: Any) {
            val protocolId = getProtocolId(packet.javaClass)
            // write protocol id to buffer
            buffer.writeShort(protocolId)
            // write packet
            getProtocol(protocolId).write(buffer, packet)
        }

        @JvmStatic
        fun read(buffer: ByteBuffer): Any {
            val protocolId = buffer.readShort()
            return getProtocol(protocolId).read(buffer)
        }
    }
}

File: protocol/src/test/scala/com/zfoo/scala/ProtocolManager.scala
package com.zfoo.scala
import com.zfoo.scala.packet.EmptyObject
import com.zfoo.scala.packet.EmptyObjectRegistration
import com.zfoo.scala.packet.VeryBigObject
import com.zfoo.scala.packet.VeryBigObjectRegistration
import com.zfoo.scala.packet.ComplexObject
import com.zfoo.scala.packet.ComplexObjectRegistration
import com.zfoo.scala.packet.NormalObject
import com.zfoo.scala.packet.NormalObjectRegistration
import com.zfoo.scala.packet.ObjectA
import com.zfoo.scala.packet.ObjectARegistration
import com.zfoo.scala.packet.ObjectB
import com.zfoo.scala.packet.ObjectBRegistration
import com.zfoo.scala.packet.SimpleObject
import com.zfoo.scala.packet.SimpleObjectRegistration
import scala.collection.mutable

object ProtocolManager {
  val MAX_PROTOCOL_NUM: Short = Short.MaxValue
  val protocols = new Array[IProtocolRegistration](MAX_PROTOCOL_NUM)
  val protocolIdMap = mutable.Map[Class[_], Short]()

  def initProtocol(): Unit = {
    // initProtocol
    protocols(0) = EmptyObjectRegistration
    protocolIdMap.put(classOf[EmptyObject], 0)
    protocols(1) = VeryBigObjectRegistration
    protocolIdMap.put(classOf[VeryBigObject], 1)
    protocols(100) = ComplexObjectRegistration
    protocolIdMap.put(classOf[ComplexObject], 100)
    protocols(101) = NormalObjectRegistration
    protocolIdMap.put(classOf[NormalObject], 101)
    protocols(102) = ObjectARegistration
    protocolIdMap.put(classOf[ObjectA], 102)
    protocols(103) = ObjectBRegistration
    protocolIdMap.put(classOf[ObjectB], 103)
    protocols(104) = SimpleObjectRegistration
    protocolIdMap.put(classOf[SimpleObject], 104)
  }

  def getProtocolId(clazz: Class[_]): Short = protocolIdMap.getOrElse(clazz, -1)

  def getProtocol(protocolId: Short): IProtocolRegistration = {
    val protocol = protocols(protocolId)
    if (protocol == null) throw new RuntimeException("[protocolId:" + protocolId + "] not exist")
    protocol
  }

  def write(buffer: ByteBuffer, packet: Any): Unit = {
    val protocolId = getProtocolId(packet.getClass)
    // write protocol id to buffer
    buffer.writeShort(protocolId)
    // write packet
    getProtocol(protocolId).write(buffer, packet)
  }

  def read(buffer: ByteBuffer): Any = {
    val protocolId = buffer.readShort
    getProtocol(protocolId).read(buffer)
  }
}